tests/dragonfly/acl_family_test.py

```diff
@@ -635,3 +645,83 @@ async def test_auth_resp3_bug(df_factory):
     }
 
     await client.close()
+
+
+@pytest.mark.asyncio
+async def test_acl_pub_sub_auth(df_factory):
+    df = df_factory.create()
+    df.start()
+    client = df.client()
+    await client.execute_command("ACL SETUSER kostas on >tmp +subscribe +psubscribe &f*o &bar")
+    assert await client.execute_command("AUTH kostas tmp") == "OK"
+
+    res = await client.execute_command("SUBSCRIBE bar")
+    assert res == ["subscribe", "bar", 1]
+
+    res = await client.execute_command("SUBSCRIBE foo")
+    assert res == ["subscribe", "foo", 2]
+
+    with pytest.raises(redis.exceptions.NoPermissionError):
+        res = await client.execute_command("SUBSCRIBE my_channel")
+
+    # PSUBSCRIBE only matches pure literals, no asterisks
+    with pytest.raises(redis.exceptions.NoPermissionError):
+        res = await client.execute_command("PSUBSCRIBE foo")
+
+    # my_channel is not in our list so the command should fail
+    with pytest.raises(redis.exceptions.NoPermissionError):
+        res = await client.execute_command("PSUBSCRIBE bar my_channel")
+
+    res = await client.execute_command("PSUBSCRIBE bar")
+    assert res == ["psubscribe", "bar", 3]
+
+
+@pytest.mark.asyncio
+async def test_acl_revoke_pub_sub_while_subscribed(df_factory):
+    df = df_factory.create()
+    df.start()
+    publisher = df.client()
+
+    async def publish_worker(client):
+        for i in range(0, 10):
+            await client.publish("channel", "message")
+
+    async def subscribe_worker(channel: aioredis.client.PubSub):
+        total_msgs = 1
+        async with async_timeout.timeout(10):
+            while total_msgs != 10:
+                res = await channel.get_message(ignore_subscribe_messages=True, timeout=5)
+                if total_msgs is not None:
+                    total_msgs = total_msgs + 1
+
+    await publisher.execute_command("ACL SETUSER kostas >tmp ON +@slow +SUBSCRIBE allchannels")
+
+    subscriber = aioredis.Redis(
+        username="kostas", password="tmp", port=df.port, decode_responses=True
+    )
+    subscriber_obj = subscriber.pubsub()
+    await subscriber_obj.subscribe("channel")
+
+    subscribe_task = asyncio.create_task(subscribe_worker(subscriber_obj))
+    await publish_worker(publisher)
+    await subscribe_task
+
+    subscribe_task = asyncio.create_task(subscribe_worker(subscriber_obj))
+    # Already subscribed, we should still be able to receive messages on channel
+    # We should not be able to unsubscribe
+    await publisher.execute_command("ACL SETUSER kostas -SUBSCRIBE -UNSUBSCRIBE")
+    await publish_worker(publisher)
+    await subscribe_task
+    # unsubscribe is not marked async and it's such a mess that it throws the error
+    # once we try to resubscribe. Instead I use the raw execute command to check that
+    # permission changes work
```

Comment:
@dranikpg you are our python guy, if you can explain to me why `unsubscribe` is not marked `async` and returns an `Awaitable` which you can't really await (forcing the exception to be thrown later) I am buying you a beer :beer:

Reply:
> if you can explain to me why unsubscribe is not marked async

Because it does `return self.execute_command("UNSUBSCRIBE", *parsed_args)` which is a coroutine

> And why the hell subscribe is marked async but not an Awaitable then

Because the `Awaitable` part just just a type hint for the interface, with an async function it's implicitly an awaitable

>  which you can't really await 

You should be able to await it freely

> Is this a common pattern in python ?

More or less and not only in python. I.e. consider

```py
async def get_value():
  return await compute_value()
```

has exactly the same behaviour as

```py
def get_value()
  return compute_value()
```

because

```py
get_value() # does nothing
await get_value() # calls into compute_value
```

as the state machine that is generated has only one state: awaiting compute_value